src/peft/tuners/ia3.py

```diff
@@ -0,0 +1,560 @@
+                f"Please check the target modules and try again."
+            )
+
+    @staticmethod
+    def _is_valid_match(key: str, target_key: str):
+        """
+        Helper function to match module names target_key and key. Makes sure that either the key is exactly the
+        target_key or the target_key is a submodule of key
+        """
+        if key.endswith(target_key):
+            if len(key) > len(target_key):
+                return key.endswith("." + target_key)  # must be a sub module
+            return True
+        return False
+
+    def _replace_module(self, parent_module, child_name, new_module, old_module):
+        setattr(parent_module, child_name, new_module)
+        new_module.weight = old_module.weight
+        if old_module.bias is not None:
+            new_module.bias = old_module.bias
+        if getattr(old_module, "state", None) is not None:
+            new_module.state = old_module.state
+            new_module.to(old_module.weight.device)
+
+        # dispatch to correct device
+        for name, module in new_module.named_modules():
+            if "ia3_" in name:
+                module.to(old_module.weight.device)
+
+    def __getattr__(self, name: str):
+        """Forward missing attributes to the wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.model, name)
+
+    def get_peft_config_as_dict(self, inference: bool = False):
+        config_dict = {}
+        for key, value in self.peft_config.items():
+            config = {k: v.value if isinstance(v, Enum) else v for k, v in asdict(value).items()}
+            if inference:
+                config["inference_mode"] = True
+        config_dict[key] = config
+        return config
+
+    def _set_adapter_layers(self, enabled=True):
+        for module in self.model.modules():
+            if isinstance(module, IA3Layer):
+                module.disable_adapters = False if enabled else True
+
+    def enable_adapter_layers(self):
+        self._set_adapter_layers(enabled=True)
+
+    def disable_adapter_layers(self):
+        self._set_adapter_layers(enabled=False)
+
+    def set_adapter(self, adapter_name):
+        for module in self.model.modules():
+            if isinstance(module, IA3Layer):
+                if module.merged:
+                    warnings.warn("Adapter cannot be set when the model is merged. Unmerging the model first.")
+                    module.unmerge()
+                module.active_adapter = adapter_name
+
+    def merge_adapter(self):
+        for module in self.model.modules():
+            if isinstance(module, IA3Layer):
+                module.merge()
+
+    def unmerge_adapter(self):
+        for module in self.model.modules():
+            if isinstance(module, IA3Layer):
+                module.unmerge()
```

Comment:
Maybe I am wrong but it seems these methods are never used? the maintained method seems to be `merge_and_unload` and users should use this method rather than `merge_adapter` and `unmerge_adapter`. cc @pacman100

Reply:
You're right! These are methods I'd copied over from LoRA, but they seem to be never used in LoRA as well. I'm removing them from IA3 for now.

Reply:
Awesome thank you !